sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -585,10 +604,23 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
   }
 }
 
-case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+case class Contains(left: Expression, right: Expression) extends StringPredicate
+  with CollationCompatibilityCheck {
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val collationId = checkCollationIds()
```

Comment:
collation id should be checked only once during analysis, not per row. Please check `Expression.checkInputDataTypes`

Reply:
fixed!